### PR TITLE
fix(platform): Separate PR reuse from the branch PR search

### LINKED ISSUE #27960: getBranchPr() reopens PR
### Describe the proposed change(s).

Discussed in https://github.com/renovatebot/renovate/discussions/27848

getBranchPr() reopens autoclosed PRs even if that PR needs approval or passing status checks, etc.

Example log:

```
DEBUG: Recreated autoclosed branch renovate/eslint-plugin-tailwindcss-3.x with sha 59eed99ccdbd709c14684e8bc18fbaf682229efd
INFO: Successfully reopened autoclosed PR
{
  "branchName": "renovate/eslint-plugin-tailwindcss-3.x"
  "title": "chore(deps): update dependency eslint-plugin-tailwindcss to v3.14.3"
  "number": 220
}
```

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Add optional `tryReuseBranchPr()` method to the platform API
- Create implementation for GitHub by extracting it from `getBranchPr()`

## Context

- Closes: #27960

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/repository/update/pr/pr-reuse.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+export async function tryReuseBranchPr(branchName: string): Promise<Pr | null> {
+  if (!platform.tryReuseBranchPr) {
+    return null;
+  }
+
+  const autoclosedPr = await platform.findPr({ branchName, state: 'closed' });
+  if (!autoclosedPr) {
+    return null;
+  }
+  logger.debug({ autoclosedPr }, 'Found autoclosed PR for branch');
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/workers/repository/update/pr/pr-reuse.ts]: Moved it

File: lib/modules/platform/types.ts [hunk 2]
```diff
@@ -274,6 +274,7 @@ export interface Platform {
     internalChecksAsSuccess: boolean,
   ): Promise<BranchStatus>;
   getBranchPr(branchName: string, targetBranch?: string): Promise<Pr | null>;
+  tryReuseBranchPr?(pr: Pr, branchName: string): Promise<Pr | null>;
```
Comment (rarkins) [inline lib/modules/platform/types.ts:277]:
```suggestion
  tryReuseAutoclosedPr?(pr: Pr, branchName: string): Promise<Pr | null>;
```

File: lib/workers/repository/update/pr/pr-reuse.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    .negate()
+    .toMillis();
+  if (closedMillisAgo > REOPEN_THRESHOLD_MILLIS) {
+    logger.debug(
+      `Found autoclosed PR ${autoclosedPr.number} but it is too old to reopen`,
+    );
+    return null;
+  }
+
+  logger.debug({ autoclosedPr }, 'Found autoclosed PR for branch');
```
Comment (rarkins) [inline lib/workers/repository/update/pr/pr-reuse.ts:41]: Inline the ID here and skip
Metadata

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by viceice — DISMISSED** [review]
we should implement the new feature for all possible platforms. especially for Gitea / Forgejo 🥰

File: lib/modules/platform/github/index.ts [hunk 4]
```diff
@@ -942,65 +939,52 @@ export async function getBranchPr(branchName: string): Promise<GhPr | null> {
...
-          body: {
-            state: 'open',
-            title,
-          },
+  return null;
+}
+
+export async function tryReuseAutoclosedPr(
+  autoclosedPr: Pr,
+  branchName: string,
```
Comment (rarkins) [inline lib/modules/platform/github/index.ts:952]: Wouldn't this be `autoclosedPr.sourceBranch` ?

Comment (zharinov) [reply lib/modules/platform/github/index.ts:952]: Well, yeah, I think so

**Review by rarkins — APPROVED** [review]

**Comment by rarkins** [comment]
LGTM, but please confirm that reopening still works as expected when needed

**Review by viceice — APPROVED** [review]

**Pull request merged**